Last modified April 15, 2021 by Luis Valdivia

In [1]:
import math

# 1.

# a.

We start by defining the function we will be working with in this problem. We can rewrite $u'(t) + 3tu+u^2 = 0$ as $u'(t) = -3tu - u^2$.

In [2]:
# INPUT: u and t
# OUTPUT: u'(t)

def f1(t,u):
    return (-u**2 - 3*t*u)

Then we define a function to compute the Forward Euler Method.

In [3]:
# INPUT: t_0 as a, t_N as b, interval length as h,
#        y = u(t_0), and f as the relevant function 
# OUTPUT: u(b)

def fem(a, b, h, y, f):
    for i in range(int((b-a)/h)):
        y += h * f(a,y)
        a += h
    return y

Given that, we can compute $u(2)$ for $h = 0.1, 0.05$ and $0.025$, done below, and store those values. 

In [4]:
fem_u0 = fem(1,2,0.1,0.5,f1)
fem_u1 = fem(1,2,0.05,0.5,f1)
fem_u2 = fem(1,2,0.025,0.5,f1)
print(fem_u0)
print(fem_u1)
print(fem_u2)

0.0011949736823790194
0.00278316214181391
0.0037820745921232626


We get that $u^{0.1}(2) = 0.0011949736823790194$, $u^{0.05}(2) = 0.00278316214181391$, and $u^{0.025}(2) = 0.0037820745921232626$. Then we can compute the ratio $r = \frac{u^{0.1}(2) - u^{0.05}(2)}{u^{0.05}(2) - u^{0.025}(2)}$ below.

In [5]:
(fem_u0 - fem_u1)/(fem_u1 - fem_u2)

1.5899175738004323

Thus, in this case $r = 1.5899175738004323$.

# b.

Now we define a function to compute the Backward Euler Method.

In [6]:
# INPUT: t_0 as a, t_N as b, interval length as h,
#        y = u(t_0), and f as the relevant function 
# OUTPUT: u(b)

def bem(a, b, h, y, f):
    for i in range(int((b-a)/h)):
        a += h
        z = y
        # newton method for finding zero
        while(abs(y + h*f(a,z) - z) > 0.0000000001):
            # rough approximation of derivative
            fp = (f(a,z+0.00000000001)-f(a,z))/0.00000000001
            z += -(y + h*f(a,z) - z)/(h*fp - 1)
        y = z
    return y

Given that, we can compute $u(2)$ for $h = 0.1, 0.05$ and $0.025$, done below, and store those values. 

In [7]:
bem_u0 = bem(1,2,0.1,0.5,f1)
bem_u1 = bem(1,2,0.05,0.5,f1)
bem_u2 = bem(1,2,0.025,0.5,f1)
print(bem_u0)
print(bem_u1)
print(bem_u2)

0.010260252143696056
0.007422732722068616
0.006115369736642983


We get that $u^{0.1}(2) = 0.010260252143696056$, $u^{0.05}(2) = 0.007422732722068616$, and $u^{0.025}(2) = 0.006115369736642983$. Then we can compute the ratio $r = \frac{u^{0.1}(2) - u^{0.05}(2)}{u^{0.05}(2) - u^{0.025}(2)}$ below.

In [8]:
(bem_u0 - bem_u1)/(bem_u1 - bem_u2)

2.1704143786078203

Thus, in this case $r = 2.1704143786078203$.

# c.

Now we define a function to compute the Trapzoidal Method.

In [9]:
# INPUT: t_0 as a, t_N as b, interval length as h,
#        y = u(t_0), and f as the relevant function 
# OUTPUT: u(b)

def tm(a, b, h, y, f):
    for i in range(int((b-a)/h)):
        a += h
        z = y
        # newton method for finding zero
        while(abs(y + 0.5*h*(f(a,z)+f(a-h,y)) - z) > 0.0000000001):
            # rough approximation of derivative
            fp = (f(a,z+0.00000000001)-f(a,z))/0.00000000001
            z += -(y + 0.5*h*(f(a,z)+f(a-h,y)) - z)/(0.5*h*(fp) - 1)
        y = z
    return y

Given that, we can compute $u(2)$ for $h = 0.1, 0.05$ and $0.025$, done below, and store those values. 

In [10]:
tm_u0 = tm(1,2,0.1,0.5,f1)
tm_u1 = tm(1,2,0.05,0.5,f1)
tm_u2 = tm(1,2,0.025,0.5,f1)
print(tm_u0)
print(tm_u1)
print(tm_u2)

0.004605087844431315
0.004824341360767579
0.004879209473654786


We get that $u^{0.1}(2) = 0.004605087844431315$, $u^{0.05}(2) = 0.004824341360767579$, and $u^{0.025}(2) = 0.004879209473654786$. Then we can compute the ratio $r = \frac{u^{0.1}(2) - u^{0.05}(2)}{u^{0.05}(2) - u^{0.025}(2)}$ below.

In [11]:
(tm_u0 - tm_u1)/(tm_u1 - tm_u2)

3.996009791460963

Thus, in this case $r = 3.996009791460963$.

# d.

Now we define a function to compute the Leapfrog Method, using Forward Euler Method to compute $U^1$.

In [12]:
# INPUT: t_0 as a, t_N as b, interval length as h,
#        y = u(t_0), and f as the relevant function 
# OUTPUT: u(b)

def lm(a, b, h, y, f):
    yn = y + h * f(a,y)
    for i in range(int((b-a)/h)):
        a += h
        t = yn
        yn = y + 2*h*f(a,yn)
        y = t 
    return y

Given that, we can compute $u(2)$ for $h = 0.1, 0.05$ and $0.025$, done below, and store those values. 

In [13]:
lm_u0 = lm(1,2,0.1,0.5,f1)
lm_u1 = lm(1,2,0.05,0.5,f1)
lm_u2 = lm(1,2,0.025,0.5,f1)
print(lm_u0)
print(lm_u1)
print(lm_u2)

1.083054426084976
0.36195906522713717
0.10168653098915817


We get that $u^{0.1}(2) = 1.083054426084976$, $u^{0.05}(2) = 0.36195906522713717$, and $u^{0.025}(2) = 0.10168653098915817$. Then we can compute the ratio $r = \frac{u^{0.1}(2) - u^{0.05}(2)}{u^{0.05}(2) - u^{0.025}(2)}$ below.

In [14]:
(lm_u0 - lm_u1)/(lm_u1 - lm_u2)

2.7705395921588423

Thus, in this case $r= 2.7705395921588423$.

# 2.

# a. 

We compute the exact solution. We can rewrite $u'=-e^{-t}u$ as $\frac{1}{u} u' = -e^{-t}$, which we can then integrate with respect to $t$ to get $\ln u = e^{-t} + C$. This is equivalent to $u(t) = e^{e^{-t} + C}$. Applying the initial condition we get $1 = e^{1 + C}$, which implies $C = -1$. 

Thus the exact solution is $u(t) = e^{e^{-t} - 1}$.

# b.

Now we define the new function we will be working with.

In [15]:
def f2(t,u):
    return (-math.exp(-t)*u)

We now use our function from part 1 to compute u(1) using the Forward Euler Method. 

In [16]:
fem_u02 = fem(0,1,0.1,1,f2)
fem_u12 = fem(0,1,0.05,1,f2)
fem_u22 = fem(0,1,0.025,1,f2)
print(fem_u02)
print(fem_u12)
print(fem_u22)

0.5018743391386775
0.5170033210793948
0.524313815571169


We get that $u^{0.1}(1) = 0.5018743391386775$, $u^{0.05}(1) = 0.5170033210793948$, and $u^{0.025}(1) = 0.524313815571169$. Then we try using the Backward Euler Method.

In [17]:
bem_u02 = bem(0,1,0.1,1,f2)
bem_u12 = bem(0,1,0.05,1,f2)
bem_u22 = bem(0,1,0.025,1,f2)
print(bem_u02)
print(bem_u12)
print(bem_u22)

0.558571554486601
0.5453048612837226
0.5384587191047242


We get that $u^{0.1}(1) = 0.558571554486601$, $u^{0.05}(1) = 0.5453048612837226$, and $u^{0.025}(1) = 0.5384587191047242$. Then we try using the Trapzoidal Method.

In [18]:
tm_u02 = tm(0,1,0.1,1,f2)
tm_u12 = tm(0,1,0.05,1,f2)
tm_u22 = tm(0,1,0.025,1,f2)
print(tm_u02)
print(tm_u12)
print(tm_u22)

0.5304679252830976
0.5312148768848872
0.5314014348251617


We get that $u^{0.1}(1) = 0.5304679252830976$, $u^{0.05}(1) = 0.5312148768848872$, and $u^{0.025}(1) = 0.5314014348251617$. Then we try using the Leapfrog Method.

In [19]:
lm_u02 = lm(0,1,0.1,1,f2)
lm_u12 = lm(0,1,0.05,1,f2)
lm_u22 = lm(0,1,0.025,1,f2)
print(lm_u02)
print(lm_u12)
print(lm_u22)

0.5400710839050569
0.5336401499626631
0.5320093232968119


We get that $u^{0.1}(1) = 0.5400710839050569$, $u^{0.05}(1) = 0.5336401499626631$, and $u^{0.025}(1) = 0.5320093232968119$. 

# c.

Now we compute $r_1 = \frac{u^{0.1}(1) - u(1)}{u^{0.05}(1) - u(1)}$ for the Forward Euler Method.

In [20]:
(fem_u02 - math.exp(math.exp(-1)-1))/(fem_u12 - math.exp(math.exp(-1)-1))

2.046243740391924

Now we compute $r_2 = \frac{u^{0.05}(1) - u(1)}{u^{0.025}(1) - u(1)}$ for the Forward Euler Method.

In [21]:
(fem_u12 - math.exp(math.exp(-1)-1))/(fem_u22 - math.exp(math.exp(-1)-1))

2.022476839246415

Now we compute $r_1 = \frac{u^{0.1}(1) - u(1)}{u^{0.05}(1) - u(1)}$ for the Backward Euler Method.

In [22]:
(bem_u02 - math.exp(math.exp(-1)-1))/(bem_u12 - math.exp(math.exp(-1)-1))

1.9584891213268731

Now we compute $r_2 = \frac{u^{0.05}(1) - u(1)}{u^{0.025}(1) - u(1)}$ for the Backward Euler Method.

In [23]:
(bem_u12 - math.exp(math.exp(-1)-1))/(bem_u22 - math.exp(math.exp(-1)-1))

1.9787034857311112

Now we compute $r_1 = \frac{u^{0.1}(1) - u(1)}{u^{0.05}(1) - u(1)}$ for the Trapezodial Method.

In [24]:
(tm_u02 - math.exp(math.exp(-1)-1))/(tm_u12 - math.exp(math.exp(-1)-1))

4.003080051538336

Now we compute $r_2 = \frac{u^{0.05}(1) - u(1)}{u^{0.025}(1) - u(1)}$ for the Trapezodial Method.

In [25]:
(tm_u12 - math.exp(math.exp(-1)-1))/(tm_u22 - math.exp(math.exp(-1)-1))

4.000744016323207

Now we compute $r_1 = \frac{u^{0.1}(1) - u(1)}{u^{0.05}(1) - u(1)}$ for the Leapfrog Method.

In [26]:
(lm_u02 - math.exp(math.exp(-1)-1))/(lm_u12 - math.exp(math.exp(-1)-1))

3.954652991335588

Now we compute $r_2 = \frac{u^{0.05}(1) - u(1)}{u^{0.025}(1) - u(1)}$ for the Leapfrog Method.

In [27]:
(lm_u12 - math.exp(math.exp(-1)-1))/(lm_u22 - math.exp(math.exp(-1)-1))

3.9884059793171844

These ratios show how fast these methods converge towards the exact solution. Both Euler Methods are about 2 because their convergence is linear, whereas Trapezoidal and Leapfrog are close to 4 because they are quadratic. These are similar to the results we got in part 1, but they differ because those in part 1 were less accurate($r = 1.5899175738004323$ for the Forward Euler Method while $r_1$ and $r_2$ are significantly closer to 2). As we get closer to the solution, we converge towards either 2 or 4, depending on which method.